File: apps/ios/Tests/CrossyUITests/CameraFollowTests.swift
import CoreGraphics
import XCTest

@testable import CrossyUI

// Camera follow (I2c): a jump that lands the cursor off-screen (a clue-browser
// tap, a Tab past the viewport edge) pans the MINIMAL distance that shows the
// cell with breathing room; a visible cursor moves nothing. The target math is
// pure; the glide is the view's hand-stepped interpolation over
// `interpolated(to:fraction:)`.

final class CameraFollowTests: XCTestCase {
    private let viewport = CGSize(width: 393, height: 560)
    private let rows = 25
    private let cols = 25

    /// Scale 1: a 25x25 board is 900x900 points, well past the viewport.
    private var camera: GridCamera {
        GridCamera(scale: 1, offset: .zero)
            .clamped(viewport: viewport, rows: rows, cols: cols)
    }

    func test_visibleCellNeedsNoPan() {
        // Cell (2, 2) sits at 72..108 on both axes: inside the margin-inset
        // viewport, so follow returns nil and the caller animates nothing.
        XCTAssertNil(
            camera.following(cell: 2 * cols + 2, viewport: viewport, rows: rows, cols: cols))
    }

    func test_offscreenRight_pansTheMinimalDistance() {
        // Cell (12, 20): x 720..756, far past the 393-point viewport; y 432..468,
        // already visible. Only x moves, and exactly far enough that the cell's
        // right edge sits one margin inside the viewport.
        let target = camera.following(
            cell: 12 * cols + 20, viewport: viewport, rows: rows, cols: cols)
        XCTAssertNotNil(target)
        XCTAssertEqual(target!.offset.x, viewport.width - GridCamera.followMarginPoints - 756)
        XCTAssertEqual(target!.offset.y, 0)
        XCTAssertEqual(target!.scale, 1, "follow is a pan, never a zoom")
    }

    func test_offscreenBelow_pansOnlyTheVerticalAxis() {
        // Cell (20, 3): y 720..756 past the 560-point viewport; x visible.
        let target = camera.following(
            cell: 20 * cols + 3, viewport: viewport, rows: rows, cols: cols)
        XCTAssertNotNil(target)
        XCTAssertEqual(target!.offset.x, 0)
        XCTAssertEqual(target!.offset.y, viewport.height - GridCamera.followMarginPoints - 756)
    }

    func test_cornerCell_staysInsideTheBoardClamp() {
        // The last cell: the pan lands on the clamp's edge, never past it (the
        // board cannot fly offscreen chasing a margin).
        let target = camera.following(
            cell: rows * cols - 1, viewport: viewport, rows: rows, cols: cols)
        XCTAssertNotNil(target)
        XCTAssertEqual(target!.offset.x, viewport.width - 900)
        XCTAssertEqual(target!.offset.y, viewport.height - 900)
    }

    func test_edgeCellUnderThePin_needsNoPanBecauseTheClampHoldsIt() {
        // Cell (0, 0) at offset zero misses its top-left margin, but the clamp
        // pins the board edge to the viewport edge: there is nowhere to pan, so
        // follow reports no-op rather than an equal camera.
        XCTAssertNil(camera.following(cell: 0, viewport: viewport, rows: rows, cols: cols))
    }

    func test_fittingBoard_neverPans() {
        // A 9x9 fits the viewport at its clamped scale; the clamp centers it and
        // every cell is visible.
        let small = GridCamera.initial(viewport: viewport, rows: 9, cols: 9)
        for cell in [0, 40, 80] {
            XCTAssertNil(small.following(cell: cell, viewport: viewport, rows: 9, cols: 9))
        }
    }

    // MARK: Word follow

    // Advancing to a clue frames the whole WORD, not just the landed cell (owner
    // 2026-07-12). `cursor` is the guarantee that never leaves the window; `word`
    // is the reach. A word wider than the zoomed window keeps the cursor on
    // screen and lets the word bleed past the far edge.

    func test_wordFullyVisible_needsNoPan() {
        // Across word row 5, cols 2..5 (cells 127..130): rect x 72..216, y
        // 180..216, comfortably inside the margins, so nothing moves.
        let word: Set<Int> = [127, 128, 129, 130]
        XCTAssertNil(
            camera.following(
                word: word, cursor: 127, viewport: viewport, rows: rows, cols: cols))
    }

    func test_acrossWordOffscreenRight_revealsTheWholeWord_notJustTheCursor() {
        // Across word row 5, cols 18..22 (cells 143..147): rect x 648..828, y
        // visible. The word (180 pt) fits the 393 pt window, so follow pans until
        // its far edge sits one margin inside: the whole word shows.
        let word: Set<Int> = [143, 144, 145, 146, 147]
        let target = camera.following(
            word: word, cursor: 143, viewport: viewport, rows: rows, cols: cols)
        XCTAssertNotNil(target)
        XCTAssertEqual(target!.offset.x, viewport.width - GridCamera.followMarginPoints - 828)
        XCTAssertEqual(target!.offset.y, 0)
        XCTAssertEqual(target!.scale, 1, "follow is a pan, never a zoom")
        // Contrast: single-cell follow stops at the cursor's own right edge (684),
        // leaving the rest of the word clipped off the right of the viewport.
        let cellOnly = camera.following(
            cell: 143, viewport: viewport, rows: rows, cols: cols)
        XCTAssertEqual(
            cellOnly!.offset.x, viewport.width - GridCamera.followMarginPoints - 684)
    }

    func test_downWordOffscreenBelow_pansOnlyVerticalToRevealTheWord() {
        // Down word col 3, rows 18..22 (cells 453,478,503,528,553): rect y
        // 648..828, x visible. The cross axis is one cell thick, so x holds.
        let word: Set<Int> = [453, 478, 503, 528, 553]
        let target = camera.following(
            word: word, cursor: 453, viewport: viewport, rows: rows, cols: cols)
        XCTAssertNotNil(target)
        XCTAssertEqual(target!.offset.x, 0, "the cross axis is a single cell")
        XCTAssertEqual(
            target!.offset.y, viewport.height - GridCamera.followMarginPoints - 828)
    }

    func test_wordWiderThanWindow_fallsBackToTheSmoothCursorFollow() {
        // Zoomed to the ceiling (72 pt cells), the full top row (cells 0..24) is
        // 1800 pt wide, far past the 393 pt window. A word that cannot be framed
        // must NOT try to: it follows the cursor cell exactly as the single-cell
        // follow does, so typing across it stays smooth (no per-keystroke flip).
        let zoomed = GridCamera(scale: GridCamera.maxScale, offset: .zero)
            .clamped(viewport: viewport, rows: rows, cols: cols)
        let word = Set(0..<cols)
        for cursor in [0, 5, 10, 18, 24] {
            XCTAssertEqual(
                zoomed.following(
                    word: word, cursor: cursor, viewport: viewport, rows: rows, cols: cols),
                zoomed.following(
                    cell: cursor, viewport: viewport, rows: rows, cols: cols),
                "a too-wide word follows the cursor cell, unchanged, at cursor \(cursor)")
        }
    }

    func test_typingAcrossAFramedWord_holdsTheCameraStill() {
        // The anti-spasm guarantee: once a word that fits is framed, advancing the
        // cursor through it returns nil (the target tracks the word, not the
        // cursor), so the camera does not move keystroke to keystroke.
        let word: Set<Int> = [143, 144, 145, 146, 147]
        let framed = camera.following(
            word: word, cursor: 143, viewport: viewport, rows: rows, cols: cols)!
        for cursor in word {
            XCTAssertNil(
                framed.following(
                    word: word, cursor: cursor, viewport: viewport, rows: rows, cols: cols),
                "a framed word holds still as the cursor crosses it (cursor \(cursor))")
        }
    }

    func test_emptyWord_fallsBackToTheCursorCell() {
        // An empty span (a defensive fallback) collapses to the single-cell follow.
        let byWord = camera.following(
            word: Set<Int>(), cursor: 12 * cols + 20,
            viewport: viewport, rows: rows, cols: cols)
        let byCell = camera.following(
            cell: 12 * cols + 20, viewport: viewport, rows: rows, cols: cols)
        XCTAssertEqual(byWord, byCell)
    }

    func test_interpolated_walksEndpointToEndpointClamped() {
        let start = GridCamera(scale: 1, offset: .zero)
        let end = GridCamera(scale: 1, offset: CGPoint(x: -100, y: 40))
        XCTAssertEqual(start.interpolated(to: end, fraction: 0), start)
        XCTAssertEqual(start.interpolated(to: end, fraction: 1), end)
        let mid = start.interpolated(to: end, fraction: 0.5)
        XCTAssertEqual(mid.offset.x, -50)
        XCTAssertEqual(mid.offset.y, 20)
        XCTAssertEqual(start.interpolated(to: end, fraction: 2), end, "fraction clamps")
    }
}

File: apps/android/ui/src/test/kotlin/crossy/ui/CameraFollowTests.kt
// Camera follow (I2c): a jump that lands the cursor off-screen pans the MINIMAL distance that shows
// the cell with breathing room, framing the whole WORD when it fits and tracking the cursor cell when
// the word is too wide; a visible cursor moves nothing. Twin of
// apps/ios/Tests/CrossyUITests/CameraFollowTests.swift; the target math is pure, the glide the view's
// hand-stepped interpolation. Defends the follow contract, not a numbered invariant.
package crossy.ui

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test

class CameraFollowTests {
    private val vw = 393f
    private val vh = 560f
    private val rows = 25
    private val cols = 25
    private val margin = GridCamera.FOLLOW_MARGIN

    // Scale 1: a 25x25 board is 900x900, well past the viewport.
    private val camera = GridCamera(1f, 0f, 0f).clamped(393f, 560f, 25, 25)

    @Test
    fun `a visible cell needs no pan`() {
        assertNull(camera.following(2 * cols + 2, vw, vh, rows, cols))
    }

    @Test
    fun `offscreen right pans the minimal distance`() {
        val target = camera.following(12 * cols + 20, vw, vh, rows, cols)
        assertNotNull(target)
        assertEquals(vw - margin - 756f, target!!.offsetX, 0.001f)
        assertEquals(0f, target.offsetY, 0.001f)
        assertEquals(1f, target.scale, 0.001f)
    }

    @Test
    fun `offscreen below pans only the vertical axis`() {
        val target = camera.following(20 * cols + 3, vw, vh, rows, cols)
        assertNotNull(target)
        assertEquals(0f, target!!.offsetX, 0.001f)
        assertEquals(vh - margin - 756f, target.offsetY, 0.001f)
    }

    @Test
    fun `a corner cell stays inside the board clamp`() {
        val target = camera.following(rows * cols - 1, vw, vh, rows, cols)
        assertNotNull(target)
        assertEquals(vw - 900f, target!!.offsetX, 0.001f)
        assertEquals(vh - 900f, target.offsetY, 0.001f)
    }

    @Test
    fun `an edge cell under the pin needs no pan because the clamp holds it`() {
        assertNull(camera.following(0, vw, vh, rows, cols))
    }

    @Test
    fun `a fitting board never pans`() {
        val small = GridCamera.initial(vw, vh, 9, 9)
        for (cell in listOf(0, 40, 80)) assertNull(small.following(cell, vw, vh, 9, 9))
    }

    // MARK: Word follow (advancing a clue frames the whole word, owner 2026-07-12)

    @Test
    fun `a fully visible word needs no pan`() {
        assertNull(camera.following(setOf(127, 128, 129, 130), 127, vw, vh, rows, cols))
    }

    @Test
    fun `an across word offscreen right reveals the whole word, not just the cursor`() {
        val word = setOf(143, 144, 145, 146, 147)
        val target = camera.following(word, 143, vw, vh, rows, cols)
        assertNotNull(target)
        assertEquals(vw - margin - 828f, target!!.offsetX, 0.001f)
        assertEquals(0f, target.offsetY, 0.001f)
        // Contrast: the single-cell follow stops at the cursor's own right edge (684).
        val cellOnly = camera.following(143, vw, vh, rows, cols)
        assertEquals(vw - margin - 684f, cellOnly!!.offsetX, 0.001f)
    }

    @Test
    fun `a down word offscreen below pans only vertical to reveal the word`() {
        val word = setOf(453, 478, 503, 528, 553)
        val target = camera.following(word, 453, vw, vh, rows, cols)
        assertNotNull(target)
        assertEquals(0f, target!!.offsetX, 0.001f)
        assertEquals(vh - margin - 828f, target.offsetY, 0.001f)
    }

    @Test
    fun `a word wider than the window falls back to the smooth cursor follow`() {
        val zoomed = GridCamera(GridCamera.maxScale, 0f, 0f).clamped(vw, vh, rows, cols)
        val word = (0 until cols).toSet()
        for (cursor in listOf(0, 5, 10, 18, 24)) {
            assertEquals(
                zoomed.following(cursor, vw, vh, rows, cols),
                zoomed.following(word, cursor, vw, vh, rows, cols),
            )
        }
    }

    @Test
    fun `typing across a framed word holds the camera still`() {
        val word = setOf(143, 144, 145, 146, 147)
        val framed = camera.following(word, 143, vw, vh, rows, cols)!!
        for (cursor in word) assertNull(framed.following(word, cursor, vw, vh, rows, cols))
    }

    @Test
    fun `an empty word falls back to the cursor cell`() {
        assertEquals(
            camera.following(12 * cols + 20, vw, vh, rows, cols),
            camera.following(emptySet(), 12 * cols + 20, vw, vh, rows, cols),
        )
    }

    @Test
    fun `interpolated walks endpoint to endpoint clamped`() {
        val start = GridCamera(1f, 0f, 0f)
        val end = GridCamera(1f, -100f, 40f)
        assertEquals(start, start.interpolated(end, 0f))
        assertEquals(end, start.interpolated(end, 1f))
        val mid = start.interpolated(end, 0.5f)
        assertEquals(-50f, mid.offsetX, 0.001f)
        assertEquals(20f, mid.offsetY, 0.001f)
        assertEquals(end, start.interpolated(end, 2f))
    }
}
